Update CHANGELOG, README and small code quality improvements (#220)
symfony/panther: README.md

### A Polymorphic Feline

Panther also gives you instant access to other BrowserKit-based implementations of `Client` and `Crawler`.

> Unlike Panther's native client, these alternative clients don't support JavaScript, CSS and screenshots capturing,

Reviewer: ```suggestion
Unlike Panther's native client, these alternative clients don't support JavaScript, CSS and screenshot capturing,
```

but they are **super-fast**!

Two alternative clients are available:

> * The first directly manipulates the Symfony kernel provided by `WebTestCase`, it's the fastest client available,

Reviewer: ```suggestion
* The first directly manipulates the Symfony kernel provided by `WebTestCase`. It is the fastest client available,
```

  but it is only available for Symfony apps.
* The second leverages the [Goutte](https://github.com/FriendsOfPHP/Goutte) web scraping library.

>   It is an intermediate between the Symfony's and the Panther's test clients. Goutte sends real HTTP requests, it is fast

Reviewer: ```suggestion
  It is an intermediate between Symfony's and Panther's test clients. Goutte sends real HTTP requests. It is fast
```

{
    public function testMyApp()
    {

>         $symfonyClient = static::createClient(); // A cute kitty: the Symfony's functional test tool

Reviewer: ```suggestion
        $symfonyClient = static::createClient(); // A cute kitty: Symfony's functional test tool
```

        // enjoy the same API for the 3 felines
        // $*client->request('GET', '...')

        $kernel = static::createKernel(); // If you are testing a Symfony app, you also have access to the kernel

        // ...
    }
}
```

> ### Creating Isolated Browsers to Test Apps Using [Mercure](https://mercure.rocks) or WebSockets

Reviewer: ```suggestion
### Creating Isolated Browsers to Test Apps Using [Mercure](https://mercure.rocks) or WebSocket
```

See: https://tools.ietf.org/html/rfc6455

Author: MDN uses Websockets, but ok

Reviewer: It actually says:

>  The WebSocket API (WebSockets)

Haha...

> Panther provides a convenient way to test applications with real-time capabilities which use [Mercure](https://symfony.com/doc/current/mercure.html), [WebSockets](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API)

Reviewer: ```suggestion
Panther provides a convenient way to test applications with real-time capabilities which use [Mercure](https://symfony.com/doc/current/mercure.html), [WebSocket](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API)
```

See: https://tools.ietf.org/html/rfc6455

and similar technologies.

> The `PantherTestCase::createAdditionalPantherClient()` creates additional, isolated, browsers; that can interact with each others.

Reviewer: ```suggestion
The `PantherTestCase::createAdditionalPantherClient()` creates additional, isolated browsers which can interact with each other.
```

For instance, this can be useful to test a chat application having several users connected simultaneously:

```php
<?php

use Symfony\Component\Panther\PantherTestCase;

class ChatTest extends PantherTestCase
{
    public function testChat(): void
    {
        $client1 = self::createPantherClient();
        $client1->request('GET', '/chat'); 
 
        // Connect a 2nd user using an isolated browser and say hi!
        $crawler2 = self::createAdditionalPantherClient()->request('GET', '/chat');
        $crawler2->submitForm('Post message', ['message' => 'Hi folks 👋😻']);

        // Wait for the message to be received by the first client
        $client1->waitFor('.message');

        // Symfony Assertions are always executed in the **primary** browser
        $this->assertSelectorTextContains('.message', 'Hi folks 👋😻');
    }
}
```

> ## Extra Documentations

Reviewer: ```suggestion
## Additional Documentation
```

> Since Panther implements the API of popular libraries, it already has an extensive documentation:

Reviewer: ```suggestion
Since Panther implements the API of popular libraries, it already has extensive documentation:
```

> * For the `Client` class, read [the BrowserKit's documentation](https://symfony.com/doc/current/components/browser_kit.html)

Reviewer: ```suggestion
* For the `Client` class, read [the BrowserKit documentation](https://symfony.com/doc/current/components/browser_kit.html)
```

> * For the `Crawler` class, read [the DomCrawler's documentation](https://symfony.com/doc/current/components/dom_crawler.html)

Reviewer: ```suggestion
* For the `Crawler` class, read [the DomCrawler documentation](https://symfony.com/doc/current/components/dom_crawler.html)
```

> * For Webdriver, read [the Facebook's PHP WebDriver documentation](https://github.com/facebook/php-webdriver)

Reviewer: ```suggestion
* For WebDriver, read [the Facebook PHP WebDriver documentation](https://github.com/facebook/php-webdriver)
```

### Environment Variables

The following environment variables can be set to change some Panther's behaviour:

* `PANTHER_NO_HEADLESS`: to disable browser's headless mode (will display the testing window, useful to debug)
* `PANTHER_NO_SANDBOX`: to disable [Chrome's sandboxing](https://chromium.googlesource.com/chromium/src/+/b4730a0c2773d8f6728946013eb812c6d3975bec/docs/design/sandbox.md) (unsafe, but allows to use Panther in containers)
* `PANTHER_WEB_SERVER_DIR`: to change the project's document root (default to `public/`)
* `PANTHER_CHROME_DRIVER_BINARY`: to use another `chromedriver` binary, instead of relying on the ones already provided by Panther
* `PANTHER_CHROME_ARGUMENTS`: to customize `chromedriver` arguments. You need to set `PANTHER_NO_HEADLESS` to fully customize.
* `PANTHER_WEB_SERVER_PORT`: to change the web server's port (default to `9080`)
* `PANTHER_WEB_SERVER_ROUTER`:  to use a web server router script which is run at the start of each HTTP request
* `PANTHER_EXTERNAL_BASE_URI`: to use an external web server (the PHP built-in web server will not be started)
* `PANTHER_CHROME_BINARY`: to use another `google-chrome` binary

### Accessing To Hidden Text

> According to the spec, WebDriver implementations returns only the **displayed** text by default.

Reviewer: ```suggestion
According to the spec, WebDriver implementations return only the **displayed** text by default.
```

> When you filter on a head tag (like `title`), the method `text()` returns an empty string. Use the method `html()` to get

Reviewer: ```suggestion
When you filter on a `head` tag (like `title`), the method `text()` returns an empty string. Use the method `html()` to get
```
